modules/standard/Path.chpl

```diff
@@ -286,4 +286,55 @@ proc file.realPath(): string throws {
       return false;
     }
   }
+
+/* Expands any environment variables in the path of the form `$<name>` or
+   `${<name>}` into their values.  If <name> does not exist, they are left
+   in place. Returns the path which includes these expansions.
+
+   :arg path: a string representation of a path, which may or may not include
+                   `$<name>` or `${<name>}`.
+   :type path: `string`
+
+   :return: `path`, having replaced all references to environment variables with
+                their values
+   :rtype: `string`
+*/
+  proc expandVars(path: string): string{
+    var path_expanded: string = path;
+    var env_arr = [("CHPL_HOME", CHPL_HOME),
+                   ("CHPL_AUX_FILESYS", CHPL_AUX_FILESYS),
+                   ("CHPL_TARGET_PLATFORM", CHPL_TARGET_PLATFORM),
+                   ("CHPL_HOST_PLATFORM", CHPL_HOST_PLATFORM),
+                   ("CHPL_HOST_COMPILER", CHPL_HOST_COMPILER),
+                   ("CHPL_TARGET_COMPILER", CHPL_TARGET_COMPILER),
+                   ("CHPL_TARGET_ARCH", CHPL_TARGET_ARCH),
+                   ("CHPL_LOCALE_MODEL", CHPL_LOCALE_MODEL),
+                   ("CHPL_COMM_SUBSTRATE", CHPL_COMM_SUBSTRATE),
+                   ("CHPL_COMM", CHPL_COMM),
+                   ("CHPL_GASNET_SEGMENT", CHPL_GASNET_SEGMENT),
+                   ("CHPL_TASKS", CHPL_TASKS),
+                   ("CHPL_LAUNCHER", CHPL_LAUNCHER),
+                   ("CHPL_TIMERS", CHPL_TIMERS),
+                   ("CHPL_UNWIND", CHPL_UNWIND),
+                   ("CHPL_MEM", CHPL_MEM),
+                   ("CHPL_MAKE", CHPL_MAKE),
+                   ("CHPL_ATOMICS", CHPL_ATOMICS),
+                   ("CHPL_NETWORK_ATOMICS", CHPL_NETWORK_ATOMICS),
+                   ("CHPL_GMP", CHPL_GMP),
+                   ("CHPL_HWLOC", CHPL_HWLOC),
+                   ("CHPL_REGEXP", CHPL_REGEXP),
+                   //("CHPL_WIDE_POINTERS", CHPL_WIDE_POINTERS),
```

Comment:
I think the intention of this function is to replace *any* environment variable, not just the ones Chapel knows about.  That will be slightly trickier.  I recommend using the Spawn module to get a subprocess; in that subprocess use known bash commands to get the list of environment variables and their values to create your array, and then follow the implementation you have outlined below

Unfortunately writing tests for that will be slightly more complicated - you'll need to use a .execenv file for your test to set dummy environment variables to values you know and compare their output explicitly.  An example of a .execenv file can be found at test/distributions/sungeun/dataPar2.execenv

Reply:
The commands to be used to get the list of the environment variables will be different on different platforms. For instance, the bash commands that I use might not work for a windows system. How do I tackle that?

Reply:
Chapel itself doesn't work on pure Windows, only linux-esque setups (i.e. cygwin, bash for Windows), so I'm not as worried about that - we've handled those cases in the past by saying they aren't supported yet in the comment on the function itself.

However, multiple bash versions is a good point.  Is there a command you know of that would at least be common for most linux-based systems?

Reply:
I think printenv should work. I'll need to look up how to construct the array from the rendered output.

Reply:
I think this approach has a few drawbacks:

1. The program has to create a new process for this, which is pretty high overhead for a string manipulation operation.
2. You have to parse the output of the command.  Note it's legal for an environment variable to contain newlines:

```
fortytwo@magrathea:~$ export FOO="asdf
fdsa=notavar
> 
> oops
> "
fortytwo@magrathea:~$ echo "$FOO"
asdf
fdsa=notavar

oops

fortytwo@magrathea:~$ env
...
TERM=xterm
FVWM_USERDIR=/home/fortytwo/.fvwm
FOO=asdf
fdsa=notavar

oops

HOSTDISPLAY=magrathea:0
```
I don't know how you would parse that reliably.  (It's not entirely an artificial concern -- I've seen a ``$TERMCAP`` environment variable with newlines in it, and lots of ``=``s.)

Another approach would be to use the ``Environment`` module to get an associative array of all the environment variables. 

Unfortunately, the ``Environment`` module is not yet implemented:  #5939 

An entirely different approach would be to turn the problem around.  Instead of getting a list of all environment variables and searching the string for each of them, go through the string and look for things that appear to be environment variables.  Then use ``getenv()`` to get the replacement text.

This has the drawback that you have to carefully process the input string.  And you have to decide what to do about input strings like ``"foo/${BAR${BA${DOS}}}"`` or ``"${OPEN_AND ooh squirrel"``.  (It's my understanding that the python ``os.path.expandvars()`` would be an appropriate model of behavior for such cases.  https://docs.python.org/2/library/os.path.html)

I think you can get at the C-level ``getenv()`` as is done in ``test/execflags/gbt/getenv.chpl``.  I don't know if that's appropriate for module code vs. for a test though.

Reply:
I agree with @cassella and furthermore think calling getenv is appropriate in a Chapel module. For example, the MPI module already does this. It would be nice to have an actual Env module focused on having a reasonable API for such queries.

The Chapel runtime includes a sys_getenv call that is intended to be more portable than the bare getenv, and to limit possible problems with the global errno. This function is available from Chapel in the Sys module. It is already called in LocaleModelHelpSetup.chpl which is probably included by default (it could depend on the configuration).